### PR TITLE
add TLS 1.3 support to WinHttp

### LINKED ISSUE #58587: WinHttp tests with Tls 1.3 are failing
This showed up in attempt to add Windows 11 test queue. We never tried to run WinHttp with TLS 1.3 before. 
At this point it is not clear if this is limitation of WinHttp or if there is work we need to do at .NET.

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-58570-merge-71d5888539224bb281/System.Net.Http.WinHttpHandler.Functional.Tests/1/console.d7f8ff6e.log?sv=2019-07-07&se=2021-09-22T18%3A25%3A20Z&sr=c&sp=rl&sig=ZUf%2BvlsRaiZhT8qXiJFDFL%2FkE%2BjPMYjAMKnuOtaIDiM%3D

     
Easy to reproduce locally:

```
System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandler_SslProtocols_Http2_Test.GetAsync_AllowedSSLVersion_Succeeds(acceptedProtocol: Tls13, requestOnlyThisProtocol: True) [FAIL]
        System.Net.Http.HttpRequestException : An error occurred while sending the request.
        ---- System.Net.Http.WinHttpException : Error 87 calling SetSessionHandleTlsOptions, 'The parameter is incorrect.'.
        Stack Trace:
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs(527,0): at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
          C:\Users\test\github\wfurt-runtime\src\libraries\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs(55,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks)
          C:\Users\test\github\wfurt-runtime\src\libraries\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs(82,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks)
          C:\Users\test\github\wfurt-runtime\src\libraries\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs(49,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout)
          C:\Users\test\github\wfurt-runtime\src\libraries\Common\tests\System\Net\Http\HttpClientHandlerTest.SslProtocols.cs(135,0): at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.<>c__DisplayClass5_1.<<GetAsync_AllowedSSLVersion_Succeeds>b__2>d.MoveNext()
          --- End of stack trace from previous location ---
          C:\Users\test\github\wfurt-runtime\src\libraries\Common\tests\System\Net\Http\LoopbackServer.cs(101,0): at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options)
          C:\Users\test\github\wfurt-runtime\src\libraries\Common\tests\System\Net\Http\HttpClientHandlerTest.SslProtocols.cs(133,0): at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.GetAsync_AllowedSSLVersion_Succeeds(SslProtocols acceptedProtocol, Boolean requestOnlyThisProtocol)
          --- End of stack trace from previous location ---
          ----- Inner Stack Trace -----
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Private.CoreLib\src\System\Environment.cs(210,0): at System.Environment.get_StackTrace()
          C:\Users\test\github\wfurt-runtime\src\libraries\Common\src\System\Runtime\ExceptionServices\ExceptionStackTrace.cs(22,0): at System.Runtime.ExceptionServices.ExceptionStackTrace.AddCurrentStack(Exception exception)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpException.cs(58,0): at System.Net.Http.WinHttpException.CreateExceptionUsingError(Int32 error, String nameOfCalledFunction)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs(1196,0): at System.Net.Http.WinHttpHandler.SetSessionHandleTlsOptions(SafeWinHttpHandle sessionHandle)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs(1121,0): at System.Net.Http.WinHttpHandler.SetSessionHandleOptions(SafeWinHttpHandle sessionHandle)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs(876,0): at System.Net.Http.WinHttpHandler.EnsureSessionHandleExists(WinHttpRequestState state)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs(901,0): at System.Net.Http.WinHttpHandler.StartRequestAsync(WinHttpRequestState state)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs(38,0): at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
             at System.Net.Http.WinHttpHandler.StartRequestAsync(WinHttpRequestState state) in System.Net.Http.WinHttpHandler.dll:token 0x60000da+0x26
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs(632,0): at System.Net.Http.WinHttpHandler.<>c.<SendAsync>b__117_0(Object s)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Private.CoreLib\src\System\Threading\ExecutionContext.cs(268,0): at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
          C:\Users\test\github\wfurt-runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs(2331,0): at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
```

### PR DESCRIPTION
**Pull request by wfurt**
WinHttp now has and the flag and it seems to work right on Windows 11
```c
#define WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_3 0x00002000
```

.NET Framework 4.8 now supports Tls13 as well somebody can use it with casting `int` like the test change.
It really does not matter as long as running on supported OS and WinHttp as all the work is done by native code.

blocking #58570
fixes #58587

### CODE REVIEW

**Comment by wfurt** [comment]
BTW Here is local test run:

```
=== TEST EXECUTION SUMMARY ===
     System.Net.Http.WinHttpHandler.Functional.Tests  Total: 1134, Errors: 0, Failed: 0, Skipped: 25, Time: 122.556s

```

and 4.8 Framework

```
  ----- start Thu 09/02/2021 14:39:14.05 ===============  To repro directly: =====================================================
  pushd C:\Users\test\github\wfurt-runtime\artifacts\bin\System.Net.Http.WinHttpHandler.Functional.Tests\net48-windows-Debug\
  xunit.console.exe System.Net.Http.WinHttpHandler.Functional.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing -verbose -parallel none
  popd
  ===========================================================================================================
    Discovering: System.Net.Http.WinHttpHandler.Functional.Tests (app domain = on [no shadow copy], method display = ClassAndMethod, method display options = None)
    Discovered:  System.Net.Http.WinHttpHandler.Functional.Tests (found 171 of 293 test cases)
    Starting:    System.Net.Http.WinHttpHandler.Functional.Tests (parallel test collections = off, max threads = 4)

…

=== TEST EXECUTION SUMMARY ===
     System.Net.Http.WinHttpHandler.Functional.Tests  Total: 558, Errors: 0, Failed: 0, Skipped: 25, Time: 17.612s
  ----- end Thu 09/02/2021 14:39:33.08 ----- exit code 0 ----------------------------------------------------------

```


all tests are passing on Windows 11.

**Review by ManickaP — APPROVED** [review]
LGTM

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.SslProtocols.cs [hunk 1]
```diff
@@ -121,7 +121,7 @@ public async Task GetAsync_AllowedSSLVersion_Succeeds(SslProtocols acceptedProto
 #if !NETFRAMEWORK
                     handler.SslProtocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12 | SslProtocols.Tls13;
 #else
-                    handler.SslProtocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
+                    handler.SslProtocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12 | (SslProtocols)12288;
```
Comment (karelz) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.SslProtocols.cs:124]:
```suggestion
                    handler.SslProtocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12 | (SslProtocols)0x3000;
```

File: src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 2]
```diff
@@ -1156,6 +1156,7 @@ private void SetSessionHandleConnectionOptions(SafeWinHttpHandle sessionHandle)
 
         private void SetSessionHandleTlsOptions(SafeWinHttpHandle sessionHandle)
         {
+            const SslProtocols Tls13 = (SslProtocols)12288; // enum is missing in .NET Standard
```
Comment (karelz) [inline src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:1159]:
```suggestion
            const SslProtocols Tls13 = (SslProtocols)0x3000; // enum is missing in .NET Standard
```

**Review by karelz — APPROVED** [review]

**Comment by karelz** [comment]
Related test failures:
```
    System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandler.GetAsync_TrailingHeaders_Ignored(includeTrailerHeader: False) [FAIL]
      System.Net.Http.HttpRequestException : An error occurred while sending the request.
      ---- System.Net.Http.WinHttpException : Error 87 calling WinHttpSetOption, 'The parameter is incorrect.'.
```
I assume we need to check OS version if the option is supported or not, prior to setting it.

- Win7 - [127 failures](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-58590-merge-270e13fcbdf8443f81/System.Net.Http.WinHttpHandler.Functional.Tests/1/console.28b8c1f1.log?sv=2019-07-07&se=2021-09-22T23%3A03%3A02Z&sr=c&sp=rl&sig=JAFhTNl1kyU8nR4JCAxwXLJmuABC8DdqtstxxX1YEn0%3D)
- Win8.1 - [129 failures](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-58590-merge-b7eb1128a74d4335b2/System.Net.Http.WinHttpHandler.Functional.Tests/1/console.3f0fcb56.log?sv=2019-07-07&se=2021-09-22T23%3A04%3A44Z&sr=c&sp=rl&sig=M1EJM1PLhyAMSnD0CWluogfdEfyzoPWXMX6mMyMFaL4%3D)
- Win10 - [137 failures](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-58590-merge-d9e389dea48e44ceb6/System.Net.Http.WinHttpHandler.Functional.Tests/1/console.892df52f.log?sv=2019-07-07&se=2021-09-22T23%3A04%3A45Z&sr=c&sp=rl&sig=pSop04En9%2F12iSj9o06T4q7DLVN0yFOmh7%2Fz%2F3qW5PU%3D)

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.SslProtocols.cs (hunk 1 above)
Comment (ManickaP) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.SslProtocols.cs:124]: I think the reason for `12288` is that's what we use in the definition:
https://github.com/dotnet/runtime/blob/9f6da559949d4f591ff87b1cef543278997d5576/src/libraries/System.Net.Primitives/ref/System.Net.Primitives.cs#L524-L524

**Comment by karelz** [comment]
Once we have it working, we should consider porting it to 6.0 -- that way we will avoid backporting it to servicing once Win11 ships officially.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.SslProtocols.cs (hunk 1 above)
Comment (karelz) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.SslProtocols.cs:124]: Ah, that is rather ugly. I wonder why? Perhaps it was auto-generated at some point?
I would still consider that hex values are more developer-friendly for reading ...

BTW: Definition is technically hex-defined:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Net.Primitives/src/System/Net/SecureProtocols/SslEnumTypes.cs#L18
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/src/Interop/Windows/SChannel/Interop.SchProtocols.cs#L31-L33

File: src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:1159]: I took the value from [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.net.securityprotocoltype?view=net-5.0#System_Net_SecurityProtocolType_Tls13). But I don't really care. It seems like the SP_PROT is platform specific mapping that happens to be the same on Windows. 
It feels like this is really matter of personal preference.

**Pull request merged**